File: app/src/main/java/me/ykrank/s1next/view/adapter/simple/SimpleViewHolderAdapter.java
package me.ykrank.s1next.view.adapter.simple;

import android.content.Context;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;

import java.util.List;

/**
 * Created by ykrank on 2016/8/1 0001.
 */
public abstract class SimpleViewHolderAdapter<D, VH extends SimpleViewHolderAdapter.BaseViewHolder> extends ArrayAdapter<D> {

    protected LayoutInflater layoutInflater;
    protected
    @LayoutRes
    int mResource;

    public SimpleViewHolderAdapter(Context context, @LayoutRes int resource, List<D> objects) {
        super(context, resource, objects);
        layoutInflater = LayoutInflater.from(context);
        this.mResource = resource;
    }

    @Override
    @SuppressWarnings("unchecked")
    public View getView(int position, View convertView, ViewGroup parent) {
        VH viewHolder;
        if (convertView == null) {
            viewHolder = onCreateViewHolder(parent);
            convertView = viewHolder.getRoot();
            convertView.setTag(viewHolder);
        } else {
            viewHolder = (VH) convertView.getTag();
        }
        onBindViewHolder(viewHolder, getItem(position), position);
        return convertView;
    }

    @Override
    public View getDropDownView(int position, View convertView, ViewGroup parent) {
        return getView(position, convertView, parent);
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }

    @Override
    public long getItemId(int position) {
        return getItem(position).hashCode();
    }

    @NonNull
    public abstract VH onCreateViewHolder(ViewGroup parent);

    public abstract void onBindViewHolder(@NonNull VH viewHolder, D data, int position);

    public static abstract class BaseViewHolder {
        private View rootView;

        public BaseViewHolder(View rootView) {
            this.rootView = rootView;
        }

        public View getRoot() {
            return rootView;
        }
    }
}

File: library/src/main/java/com/github/ykrank/androidtools/ui/adapter/simple/SimpleViewHolderAdapter.kt
package com.github.ykrank.androidtools.ui.adapter.simple

import android.content.Context
import androidx.annotation.LayoutRes
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter

/**
 * Created by ykrank on 2016/8/1 0001.
 */
abstract class SimpleViewHolderAdapter<D, VH : SimpleViewHolderAdapter.BaseViewHolder>(context: Context, @param:LayoutRes @field:LayoutRes
protected var mResource: Int, objects: List<D>) : ArrayAdapter<D>(context, mResource, objects) {

    protected var layoutInflater: LayoutInflater = LayoutInflater.from(context)

    override fun getView(position: Int, convertView_: View?, parent: ViewGroup): View {
        var convertView = convertView_
        val viewHolder: VH
        if (convertView == null) {
            viewHolder = onCreateViewHolder(parent)
            convertView = viewHolder.root
            convertView.tag = viewHolder
        } else {
            viewHolder = convertView.tag as VH
        }
        onBindViewHolder(viewHolder, getItem(position), position)
        return convertView
    }

    override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {
        return getView(position, convertView, parent)
    }

    override fun hasStableIds(): Boolean {
        return true
    }

    override fun getItemId(position: Int): Long {
        return getItem(position)?.hashCode()?.toLong() ?: position.toLong()
    }

    abstract fun onCreateViewHolder(parent: ViewGroup): VH

    abstract fun onBindViewHolder(viewHolder: VH, data: D?, position: Int)

    abstract class BaseViewHolder(val root: View)
}
